### ISSUE TITLE
Context manager support for all gears

### ISSUE THREAD

**Issue by sueskind** [abhiTronix/vidgear#309]
## Brief Description

Many Python developers (including me) expect objects with an "open/start -> close" workflow to behave as context managers, i.e. support the `with ... as ...` construct.

### Context

This feature would implement the `__enter__()` and `__exit__()` methods for all gears. Internally these only call the gear's `start()` and `close()/stop()/clean()` methods.

There are two major advantages for using Python's context managers instead of calling `start()` and `close()` independently:

- The resource is automatically closed so calling `close()` in the end cannot be forgotten.
- The resource is even closed in case of an exception since the `with` construct runs the contained code as `try ... finally`. The resource is closed in the finally block.

Furthermore:

- It will make the API consistent with other established Python modules/packages, e.g. [`open()`](https://docs.python.org/3/library/functions.html#open), [`threading.Lock`](https://docs.python.org/3/library/threading.html#using-locks-conditions-and-semaphores-in-the-with-statement) or [`pytest.raises()`](https://docs.pytest.org/en/stable/reference/reference.html#pytest.raises)
- It will be backwards compatible since it does not change existing APIs.
- There is no risk involved.
- Future usages will be more correct since the `close()` cannot be forgotten.
- (Opinion) Nicer looking code.

#### Usage example

Simplifying the [NetGear bare-minimum example](https://abhitronix.github.io/vidgear/v0.2.5-stable/gears/netgear/usage/):
```python
# import required libraries
from vidgear.gears import VideoGear
from vidgear.gears import NetGear

# open any valid video stream(for e.g `test.mp4` file)
# Define Netgear Server with default parameters
with VideoGear(source="test.mp4") as stream, NetGear() as server:
    
    # loop over until KeyBoard Interrupted
    while True:
    
        try:
    
            # read frames from stream
            frame = stream.read()
    
            # check for frame if Nonetype
            if frame is None:
                break
    
            # {do something with the frame here}
    
            # send frame to server
            server.send(frame)
    
        except KeyboardInterrupt:
            break

# video stream and server safely closed
```

### Your Current Environment
* VidGear version: 0.2.6
* Branch: testing
* Python version: Python 3.10.4
* Operating System and version: Arch Linux (Kernel 5.17.9-arch1-1)

**Comment by abhiTronix** [comment]
> Many Python developers (including me) expect objects with an "open/start -> close" workflow to behave as context managers, i.e. support the with ... as ... construct.
> There are two major advantages for using Python's context managers instead of calling start() and close() independently:
> 
> The resource is automatically closed so calling close() in the end cannot be forgotten.
> The resource is even closed in case of an exception since the with construct runs the contained code as try ... finally. The resource is closed in the finally block.

@sueskind Thanks for the proposal and related PR. But let's talk about why `start()` and `stop()/close()` are there in the first place, what are cons of removing/replacing them with Python's context managers:

### The `start()` function is there because:
- **To create internal multi-threaded video-frames extractor on choice**: Therefore, if you don't want to create multi-threads for any reason _(such as extracting meta data from YouTube stream in CamGear)_, you can skip calling and no threads will be created unnecessarily.
- **To `return self`**: Returning `self` from a method simply means that your method returns a reference to the instance API object on which it was called. 
  - This famous in object oriented APIs (like vidgear's) that are designed as a [fluent interface](https://en.wikipedia.org/wiki/Fluent_interface) that encourages [method cascading](https://en.wikipedia.org/wiki/Method_cascading). 
  - It helps in both returning and altering internal Class Parameter directly from python script.
- **To match syntax of opencv-python code**: `start()` function is analogous to `open()` function in OpenCV's python API. So that developers can easily related with vidgear's code if they're already familiar with OpenCV.

#### Cons of removing /replacing `start()` function with Context manager's `__enter__`:
- **No method present to avoid creating multiple-threads with `__enter__`**: Not everyone interested in creating them.
- **Returning `self` from `__enter__` is considered bad practice and can break things in vidgear**: Returning `self` from `__enter__` should only be done if `self` is fully initialized beforehand but most of API's in vidgear needs to be altered after initializing.
- **Will create confusion among developers about syntax**: After all Vidgear 's motto is to make video-processing easier as possible.



### The `stop()/close()` function is there because:
- **To cleanup resources on exit manually**: Resources must be cleaned manually at the end. 
  - **Why manually you may ask**, because Threading and Subprocess errors are the worst in python, and will sometimes cause background thread to freeze completely even when main thread is freed. This leads to main thread also stuck in while loop and can only be freed by pressing `CRTL+C` and therefore seperate call on `stop()/close()` function is necessary to clean up resources. 
- **To match syntax of opencv-python code**: `stop/close()` function is analogous to `release()` function in OpenCV's python API. So that developers can easily related with vidgear's code if they're already familiar with OpenCV.

#### Cons of removing /replacing `start()` function with Context manager's `__enter__`:
- `__exit__` is not executed when `__enter__` raises exception: This is bad because if system resources(such as Socket bounded to an address in NetGear and Subprocess pipeline in WriteGear) are not released, then API will fail to exit, or even bad, will fail to initialize on rerun(since resources are still allocated). 
- **With statement machinery will interpret exceptions from `__exit__` as a failure of `__exit_`_, not as a propagation of the original exception**: Therefore any error in API will not be seen in traceback.
- **Will create confusion among developers about syntax**

**Comment by abhiTronix** [comment]
@sueskind And most importantly **Python Context managers are not backward compatible with current API but current API can be converted to Python Context managers easily and safely**. Here's a crude example:

```python
from vidgear.gears import VideoGear
import cv2
import datetime
import time


class CM_VideoGear:
    def __init__(self, cam_address, stabilize=False, reset_attempts=50, reset_delay=5):
        self.cam_address = cam_address
        self.stabilize = stabilize
        self.reset_attempts = reset_attempts
        self.reset_delay = reset_delay

    def __enter__(self):
        self.source = VideoGear(
            source=self.cam_address, stabilize=self.stabilize
        ).start()
        self.running = True

    def read(self):
        if self.source is None:
            return None
        if self.running and self.reset_attempts > 0:
            frame = self.source.read()
            if frame is None:
                self.source.stop()
                self.reset_attempts -= 1
                print(
                    "Re-connection Attempt-{} occured at time:{}".format(
                        str(self.reset_attempts),
                        datetime.datetime.now().strftime("%m-%d-%Y %I:%M:%S%p"),
                    )
                )
                time.sleep(self.reset_delay)
                self.source = VideoGear(
                    source=self.cam_address, stabilize=self.stabilize
                ).start()
                # return previous frame
                return self.frame
            else:
                self.frame = frame
                return frame
        else:
            return None

    def __exit__(self):
        self.running = False
        self.reset_attempts = 0
        self.frame = None
        if not self.source is None:
            self.source.stop()


if __name__ == "__main__":
    # open any valid video stream
    with CM_VideoGear(
        cam_address="rtsp://wowzaec2demo.streamlock.net/vod/mp4:BigBuckBunny_115k.mov",
        reset_attempts=20,
        reset_delay=5,
    ) as stream:
        # loop over
        while True:

            # read frames from stream
            frame = stream.read()

            # check for frame if None-type
            if frame is None:
                break

            # {do something with the frame here}

            # Show output window
            cv2.imshow("Output", frame)

            # check for 'q' key if pressed
            key = cv2.waitKey(1) & 0xFF
            if key == ord("q"):
                break

    # close output window
    cv2.destroyAllWindows()
```

**Comment by sueskind** [comment]
@abhiTronix Thank you for your fast and elaborate response to my proposal and that you consider adding it.

First, I think there is a small misunderstanding. I don't want to **remove or replace** the old methods (`start()`/`stop()`/`close()`/...) but **add** the `__enter__` and `__exit__` methods which simply call the current API [like so](https://github.com/sueskind/vidgear/blob/ba3bfa1d7e920aef84fdfb4dc4da92f7bd48aa6a/vidgear/gears/camgear.py#L421):

```python
def __enter__(self):
    return self.start()

def __exit__(self, exc_type, exc_val, exc_tb):
    self.stop()
```

The "redundant" option to open/release a resource with the traditional `start()/stop()` methods **and** the context manager is very common (see my given examples for context managers above, they all support the redundant API).

I agree with all of your points. However, I think they all still apply after adding the context manager support.

You mentioned the similarity to the OpenCV API. There are multiple examples where users would have liked context managers in OpenCV, too ([example1](https://stackoverflow.com/questions/19086343/with-statement-to-read-from-videocapture-in-opencv), [example2](https://gist.github.com/v-goncharenko/d9e1fef1ee40fa36f610acebeabbcc25)). This is no surprise since using context managers plays a big part in writing idiomatic Python code, as seen in lots of articles on this topic or books about clean Python code:

- [Wikibooks on Python Idioms](https://en.wikibooks.org/wiki/Python_Programming/Idioms#Principles)
- [Python Hitchhiker's Guide (O'Reilly) on Code Style](https://docs.python-guide.org/writing/style/#read-from-a-file)
- [Raymond Hettinger at Pycon US](https://github.com/JeffPaine/beautiful_idiomatic_python#decorators-and-context-managers)
- [Google's Python style guide](https://google.github.io/styleguide/pyguide.html#294-decision)

**Comment by abhiTronix** [comment]
> First, I think there is a small misunderstanding. I don't want to remove or replace the old methods (start()/stop()/close()/...) but add the __enter__ and __exit__ methods which simply call the current API [like so](https://github.com/sueskind/vidgear/blob/ba3bfa1d7e920aef84fdfb4dc4da92f7bd48aa6a/vidgear/gears/camgear.py#L421):
> 
> def __enter__(self):
>     return self.start()
> 
> def __exit__(self, exc_type, exc_val, exc_tb):
>     self.stop()

@sueskind Don't you think presence of `__enter__` itself will replace the functioning of `start()`, Since now you cannot call `start()` manually as it will create duplicate threads for same resource, and the whole thing will crash. So It is definitely breaking the present API(Same for your current PR which's affected by this problem). Also same with `__exit__`. So only thing you can do it is remove `start()` and `stop/close()` entirely and place its code inside `__enter__` and `__exit__`, but that will user no choice but to create and destroy threads every single time.

> The "redundant" option to open/release a resource with the traditional start()/stop() methods and the context manager is very common (see my given examples for context managers above, they all support the redundant API).

Unless if you're not using multi-threading, which hates duplicated threads for same resource. Read about GIL and you'll know why it is bad.

> I agree with all of your points. However, I think they all still apply after adding the context manager support.
> 
> You mentioned the similarity to the OpenCV API. There are multiple examples where users would have liked context managers in OpenCV, too ([example1](https://stackoverflow.com/questions/19086343/with-statement-to-read-from-videocapture-in-opencv), [example2](https://gist.github.com/v-goncharenko/d9e1fef1ee40fa36f610acebeabbcc25)). This is no surprise since using context managers plays a big part in writing idiomatic Python code, as seen in lots of articles on this topic or books about clean Python code:
> 
> [Wikibooks on Python Idioms](https://en.wikibooks.org/wiki/Python_Programming/Idioms#Principles)
> [Python Hitchhiker's Guide (O'Reilly) on Code Style](https://docs.python-guide.org/writing/style/#read-from-a-file)
> [Raymond Hettinger at Pycon US](https://github.com/JeffPaine/beautiful_idiomatic_python#decorators-and-context-managers)
> [Google's Python style guide](https://google.github.io/styleguide/pyguide.html#294-decision)

**I'm not seeing any advantage except clear and beautiful coding format. I don't think introducing more bugs to current API for the same is a favorable thing to do in my opinion. You're free to [write a wrapper around any vidgear's API and add context manager on top of that](https://github.com/abhiTronix/vidgear/issues/309#issuecomment-1138083912), but changes to API itself is not I'm willing to do. Thank you for the PR but I'll have to close this.**

**Comment by abhiTronix** [comment]
@sueskind I'm looking into adding this Context Manager feature for non-multithreaded gears like WriteGear and StreamGear, and it will be a great idea. I'm also implementing this feature for DeFFcode APIs, so you can expect to see them in Upcoming Gears and existing Gears(except CamGear) once DeFFcode replaces OpenCV in #148.

**Comment by abhiTronix** [comment]
**Context manager support for WriteGear API implemented in commit 0d1592c5ed8708a7887a42a815a68ace3fcc104c**

**Comment by abhiTronix** [comment]
Context manager support for StreamGear API implemented in commit https://github.com/abhiTronix/vidgear/commit/7bd635fd8e304c5195da34f9f3266be126154d95
